Feature: Added the DateAndTime Picker Component

Deepanshu1230 opened the pull request:
Description
This pull request introduces a new, highly configurable DateTimePicker component to the @tambo-ai/react library. This component allows users to select a single date and time, and it is designed to be fully accessible, timezone-aware, and internationalized.

An interactive showcase page has also been created to demonstrate all component features and facilitate manual testing.

Key Features
Timezone Aware: Displays times in the user's local timezone while emitting standardized UTC ISO 8601 strings via the onChange prop.
Internationalization: Supports multiple locales for language and date formatting. The showcase demonstrates this with over 10 languages.
12/24 Hour Cycles: Includes an hourCycle prop to easily switch between 12-hour (AM/PM) and 24-hour time formats.
Date Constraints: Supports min and max props to restrict the selectable date range.
Full Keyboard Navigation: The calendar is fully accessible, allowing users to navigate and select dates using only arrow keys, Page Up/Down, Home/End, and Enter.
Props Validation: Uses Zod for robust and type-safe props schema validation.




https://github.com/user-attachments/assets/5c99246c-b0c6-470d-88a1-b98dde71a0a2






Screenshot:
<img width="1355" height="648" alt="image" src="https://github.com/user-attachments/assets/94b9e074-eeb0-4f38-b750-bd2705faca29" />
<img width="1354" height="649" alt="image" src="https://github.com/user-attachments/assets/e6b22643-2d07-4d1e-a1cb-76e04e53f34d" />
<img width="1350" height="642" alt="image" src="https://github.com/user-attachments/assets/6d81e8a9-e667-47dc-b04a-b5e00eaef471" />



How to Test
Navigate to the showcase page at /components/date-time-picker.
Verify that picking a date and time updates the "Current ISO Value" with a valid UTC ISO string.
Test the "Toggle Hour Cycle" button with both English and another language (e.g., French) to ensure the time format updates correctly in the input field.
Use the language dropdown to select different locales and confirm the calendar month/day names and the input text are translated.
Open the calendar and use only the keyboard to navigate (arrow keys, Page Up/Down) and select a date (Enter key). Confirm that focus is tracked and selection works as expected.
Click the "Clear" button and verify that the onChange callback is triggered with a null value.


Closes issue #880

Deepanshu1230 commented:
Hi @michaelmagan you can review this request.

michaelmagan commented:
@Deepanshu1230, thanks for opening this. A few notes:
1. Could you make it look better? It doesn't fit well with the style of our existing components.
2. On the showcase site, you should have a generative demo. For example, I can say "2 weeks from today" and then it will generate that component with the entered values, allowing me to interact with it.

Please update the demo once you make these changes.

Once you've done that, please let me know, and I'll provide a thorough review.

Deepanshu1230 commented:
Hi @michaelmagan ,
 * It took me around 2 days to go through and understand the whole codebase, but now I feel much more confident with it.
 * I’ve addressed all your feedback:
Improved the styling to fit the existing components.
Added the generative demo (e.g., “2 weeks from today”).
Updated the showcase site and included a demo video.

Everything should now be ready for your review.

Deepanshu1230 commented:
"Hi @michaelmagan , I’ve implemented all the requested changes (styling improvements, generative demo, updated showcase with video/screenshots). Could you please take another look when you get a chance? Thanks!"

alecf reviewed (changes requested):
Thanks for this, the picker looks nice!

There are a few key things about the tambo components that make them "special" over standard components:
1. they are installed similar to shadcn, so they have to "live" in the cli directory (see instructions below) - so ultimately `npx tambo add date-picker` should add the date picker to the current project. This will also require a config.json to get the dependencies right (like `date-fns`)
2. they need to be usable in a chat context. Since this particular component is an interactive picker, the demo should show how the AI can "ask" the user for a date and the user can "submit" a date instead of typing a response

So for the AI demo you could have a sample prompt like "Ask me what day I would like to schedule my next appointment" and the AI should respond by showing the date picker. The showcase could have a little wrapper around the picker that says "submit" (or perhaps the component itself could have a boolean prop `showSubmitButton`) that when it is pressed, the component sends the date to the AI using `sendThreadMessage` from the `useTambo` hook.

Deepanshu1230 commented on showcase/src/app/components/(generative)/date-time-picker/page.tsx:
```diff
@@ -0,0 +1,224 @@
+  "pt-BR": ptBR,
+};
+
+function parseNaturalDate(text: string): Date | null {
```
@alecf  Thanks for the review You're absolutely right! I can see how that would be confusing - the natural language parsing isn't actually part of the DateTimePicker component itself, it was just a demo feature I added to the showcase page.

akhileshrangani4 commented:
Hey @Deepanshu1230, appreciate the work you put into this. We’ve decided not to move forward with this component at the moment, so I’ll be closing this PR.
